libs/src/service-selection/ServiceSelection.test.js

```diff
@@ -0,0 +1,251 @@
+      .reply(400)
+
+    const s = new ServiceSelection({
+      getServices: () => [good, bad]
+    })
+    const service = await s.select()
+    assert.strictEqual(service, good)
+  })
+
+  it('prefers a faster service', async () => {
+    const fast = 'https://fast.audius.co'
+    nock(fast)
+      .get('/health_check')
+      .reply(200)
+
+    const slow = 'https://slow.audius.co'
+    nock(slow)
+      .get('/health_check')
+      .delay(200)
+      .reply(400)
+
+    const s = new ServiceSelection({
+      getServices: () => [fast, slow]
+    })
+    const service = await s.select()
+    assert.strictEqual(service, fast)
+  })
+
+  it('prefers a slower healthy service', async () => {
+    const fast = 'https://fast.audius.co'
+    nock(fast)
+      .get('/health_check')
+      .reply(400)
+
+    const slow = 'https://slow.audius.co'
+    nock(slow)
+      .get('/health_check')
+      .delay(200)
+      .reply(200)
+
+    const s = new ServiceSelection({
+      getServices: () => [fast, slow]
+    })
+    const service = await s.select()
+    assert.strictEqual(service, slow)
+    assert.deepStrictEqual(s.unhealthy, new Set([fast]))
+  })
+
+  it('should find the needle in the haystack', async () => {
+    // Single good service
+    const needle = 'https://needle.audius.co'
+    nock(needle)
+      .get('/health_check')
+      .reply(200)
+    // Many bad services
+    const haystack = Array.from({ length: 20 }, (v, i) => `https://${i}.audius.co`)
+    haystack.forEach(hay => {
+      nock(hay).get('./health_check').reply(400)
+    })
+
+    const s = new ServiceSelection({
+      getServices: () => [...haystack, needle],
+      maxConcurrentRequests: 2,
+      requestTimeout: 100
+    })
+    const service = await s.select()
+    assert.strictEqual(service, needle)
+  })
+
+  it('should pick null if there is no healthy service', async () => {
+    const bad1 = 'https://bad1.audius.co'
+    nock(bad1)
+      .get('/health_check')
+      .reply(400)
+
+    const bad2 = 'https://bad2.audius.co'
+    nock(bad2)
+      .get('/health_check')
+      .reply(400)
+
+    const s = new ServiceSelection({
+      getServices: () => [bad1, bad2],
+      // Short timeout otherwise, we'll wait for a long time for some request to succeed
+      // TODO: consider ammending promiseFight to early exit
+      requestTimeout: 100
+    })
+    const service = await s.select()
+    assert.strictEqual(service, null)
+  })
+
+  it('respects a whitelist', async () => {
+    const fast = 'https://fast.audius.co'
+    nock(fast)
+      .get('/health_check')
+      .reply(200)
+
+    const slow = 'https://slow.audius.co'
+    nock(slow)
+      .get('/health_check')
+      .delay(200)
+      .reply(200)
+
+    const s = new ServiceSelection({
+      getServices: () => [fast, slow],
+      whitelist: new Set([slow])
+    })
+    const service = await s.select()
+    assert.strictEqual(service, slow)
+  })
+})
+
+describe('ServiceSelection withBackupCriteria', () => {
```

Comment:
@hareeshnagaraj these tests might be a little bit more illuminating w.r.t. the "version" preference stuff

Reply:
is the 'behind: true' a boolean representing relation to current version? jw because our health checks currently don't return bool like that

Reply:
yeah this is just a test case. not a real world example, but it illustrates how something behind (or subject to other criteria) could eventually be picked